File: 166. Fraction to Recurring Decimal/Python/Fraction to Recurring Decimal.py
class Solution(object):
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        if numerator == 0:
            return '0'
        result = ''
        if (numerator > 0) != (denominator > 0):
            result = '-'
        numerator = abs(numerator)
        denominator = abs(denominator)
        result += str(numerator // denominator)
        numerator %= denominator
        if numerator == 0:
            return result
        result += '.'
        memory = {}
        memory[numerator] = len(result)
        while numerator:
            numerator *= 10
            result += str(numerator // denominator)
            numerator %= denominator
            if numerator in memory:
                start_index = memory[numerator]
                result = result[:start_index] + '(' + result[start_index:] + ')'
                break
            memory[numerator] = len(result)
        return result

File: 166. Fraction to Recurring Decimal/C++/Fraction to Recurring Decimal.cpp
class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) {
            return "0";
        }
        string result;
        if ((numerator > 0) != (denominator > 0)) {
            result = "-";
        }
        long num = abs((long) numerator);
        long den = abs((long) denominator);
        result += to_string(num / den);
        num %= den;
        if (num == 0) {
            return result;
        }
        result += ".";
        unordered_map<long, int> memory;
        memory[num] = result.length();
        while (num != 0) {
            num *= 10;
            result += to_string(num / den);
            num %= den;
            if (memory.find(num) != memory.end()) {
                int startIndex = memory[num];
                result.insert(startIndex, "(");
                result += ")";
                break;
            }
            memory[num] = result.length();
        }
        return result;
    }
};